asynciterator.ts

```diff
@@ -1252,6 +1254,135 @@ function destinationFillBuffer<S>(this: InternalSource<S>) {
 }
 
 
+export abstract class SynchronousTransformIterator<S, D = S> extends AsyncIterator<D> {
+  private _destroySource: boolean;
+  protected constructor(protected _source: AsyncIterator<S>, protected options: { destroySource?: boolean } = {}) {
+    /* eslint-disable no-use-before-define */
+    super();
+    this._destroySource = options.destroySource !== false;
+    const cleanup = () => {
+      _source.removeListener('end', onSourceEnd);
+      _source.removeListener('error', onSourceError);
+      _source.removeListener('readable', onSourceReadable);
+    };
+    const onSourceEnd = () => {
+      cleanup();
+      this.close();
+    };
+    const onSourceError = (err: Error) => {
+      this.emit('error', err);
+    };
+    const onSourceReadable = () => {
+      if (this.readable) {
+        // TODO: I'm not completely sure as to why this is needed but without
+        //       the following line, some use cases relying on flow mode (i.e.
+        //       consuming items via `on('data', (data) => {})`) do not work.
+        //       It looks like the debouncing that happens in `set readable()`
+        //       in `AsyncIterator` prevents the event from firing as `this`
+        //       is already readable.
+        this.emit('readable');
+      }
+      else {
+        this.readable = true;
+      }
+    };
+    _source.on('end', onSourceEnd);
+    _source.on('error', onSourceError);
+    _source.on('readable', onSourceReadable);
+    if (_source.done)
+      onSourceEnd();
+    else if (_source.readable)
+      onSourceReadable();
+  }
+
+  destroy(cause?: Error): void {
+    this._source.destroy(cause);
+    super.destroy(cause);
+  }
+
+  public close() {
+    if (this._destroySource)
+      this._source.destroy();
+    super.close();
+  }
+}
+
+interface Transform {
+  fn: Function,
+  next?: Transform
+}
+
+export class SyncTransformIterator<T, D = T> extends SynchronousTransformIterator<T, D> {
```

Comment:
`SyncTransformIterator` and `SynchronousTransformIterator` are too similar IMHO. Perhaps:

```
SynchronousTrasformIterator => BaseSynchronousIterator
SyncTransformIterator => SynchronousTransformIterator
```

Reply:
Let's just merge both classes.